src/domain/notifications/v2/notifications.repository.ts

```diff
@@ -88,19 +92,188 @@ export class NotificationsRepositoryV2 implements INotificationsRepositoryV2 {
+        .from(NotificationSubscription)
+        .where(
+          `chain_id = :chainId
+          AND safe_address = :safeAddress
+          AND push_notification_device.id = :deviceId
+          AND (
+            signer_address = :signerAddress OR signer_address IS NULL
+          )`,
+          {
+            chainId: safe.chainId,
+            safeAddress: safe.address,
+            deviceId: args.deviceId,
+            signerAddress: args.signerAddress ?? null,
+          },
+        )
+        .execute();
+    }
+  }
+
+  private async upsertSubscription(args: {
+    authPayload: AuthPayload;
+    upsertSubscriptionsDto: UpsertSubscriptionsDto;
+    deviceId: number;
+  }): Promise<Array<NotificationSubscription>> {
+    const subscriptionsToInsert: Partial<NotificationSubscription>[] = [];
+    for (const safe of args.upsertSubscriptionsDto.safes) {
+      const device = new NotificationDevice();
+      device.id = args.deviceId;
+      subscriptionsToInsert.push({
+        chain_id: safe.chainId,
+        safe_address: safe.address,
+        signer_address: args.authPayload.signer_address ?? null,
+        push_notification_device: device,
+      });
+    }
+
+    const databaseTransaction =
+      this.postgresDatabaseService.getTransactionRunner();
+
+    const insertResult = await databaseTransaction.upsert(
+      NotificationSubscription,
+      subscriptionsToInsert,
+      [
+        'chain_id',
+        'safe_address',
+        'signer_address',
+        'push_notification_device',
+      ],
+    );
+    const subscriptionIds: Array<number> = insertResult.identifiers.map(
+      (subscriptionIdentifier) => subscriptionIdentifier.id,
+    );
+
+    const subscriptions = await this.getSubscriptionsById(subscriptionIds);
+
+    return subscriptions;
+  }
+
+  public async getSubscriptionsById(
+    subscriptionIds: Array<number>,
+  ): Promise<Array<NotificationSubscription>> {
+    const databaseTransaction =
+      this.postgresDatabaseService.getTransactionRunner();
+
+    return await databaseTransaction.find(NotificationSubscription, {
+      where: { id: In(subscriptionIds) },
     });
   }
 
-  getSafeSubscription(args: {
+  private async insertSubscriptionNotificationTypes(arg: {
+    upsertSubscriptionsDto: UpsertSubscriptionsDto;
+    subscriptions: Array<NotificationSubscription>;
+  }): Promise<void> {
+    const notificationTypesMap = new Map<string, NotificationType>(); // A map of all the notification types in request along with their database entity
+    const notificationTypes = arg.upsertSubscriptionsDto.safes.flatMap(
+      (safe) => safe.notificationTypes,
+    );
+    const uniqueNotificationTypes = new Set(notificationTypes);
+    const databaseTransaction =
+      this.postgresDatabaseService.getTransactionRunner();
+
+    const notificationTypeObjects = await databaseTransaction.find(
+      NotificationType,
+      {
+        where: { name: In([...uniqueNotificationTypes]) },
+      },
+    );
```

Comment:
“Yes, that’s correct. This is a simple select query, so I wouldn’t be too concerned about caching it specially that the `name` column is also indexed. However, we can go ahead and cache it since the data doesn’t change often.

Reply:
Since the PR is approved, I'm gonna resolve this thread.